<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • minSNPs</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">minSNPs</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/MinSNPs_Workflow.html">MinSNPs Workflow</a>
    </li>
    <li>
      <a href="../articles/user_manual.html">MinSNPs User Manual</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/ludwigHoon/minSNPs/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/ludwigHoon/minSNPs/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.3" id="minsnps-v003">minSNPs v0.0.3<small>2022-08-27</small><a class="anchor" aria-label="anchor" href="#minsnps-v003"></a></h2>
<p>Minor changes &amp; bug removal</p>
<div class="section level3">
<h3 id="changes-0-0-3">Changes<a class="anchor" aria-label="anchor" href="#changes-0-0-3"></a></h3>
<ul><li>
<code>find_optimised_snps</code> can use <code><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">BiocParallel::SnowParam()</a></code> for windows.</li></ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.2" id="minsnps-v002">minSNPs v0.0.2<small>2022-03-28</small><a class="anchor" aria-label="anchor" href="#minsnps-v002"></a></h2>
<p>Minor changes &amp; testing</p>
<div class="section level3">
<h3 id="changes-0-0-2">Changes<a class="anchor" aria-label="anchor" href="#changes-0-0-2"></a></h3>
<ul><li>
<code>find_optimised_snps</code> can receive an additional argument <code>output_progress = TRUE</code>, such that progress can be shown in terminal when looking for multiple result.</li>
<li>
<code>write_fasta</code> bug where only the first letter is written to the file is fixed.</li>
<li>
<code>output_result</code> bug where residuals are not returned.</li>
<li>
<code>process_allele</code> will print names of duplicated isolate, and keep only the first instance of the found isolate.</li>
<li>
<code>iterate_merge</code> can be used to merge alignment matrices.</li>
<li>
<code>find_optimised_snps</code> reduced memory usage.</li>
<li>
<code>resolve_IUPAC_missing</code> can be used to substitute the ambiguity codes found in the sequence.</li>
<li>
<code>output_result</code> updated such that it can be presented correctly in excel.</li>
<li>
<code>cal_fn</code> and <code>cal_fp</code> can be used to calculate and identify false negatives and false postives respectively. To be called with <code>cal_met_snp</code>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.0.1" id="minsnps-v001">minSNPs v0.0.1<small>2021-09-08</small><a class="anchor" aria-label="anchor" href="#minsnps-v001"></a></h2>
<p>First release to CRAN.</p>
<div class="section level3">
<h3 id="changes-0-0-1">Changes<a class="anchor" aria-label="anchor" href="#changes-0-0-1"></a></h3>
<ul><li>
<code>read_fasta</code> will be used to read fasta file, the sequence name can contain spaces.</li>
<li>
<code>write_fasta</code> will be used to write fasta file, the sequence name can contain spaces.</li>
<li>
<code>process_allele</code> can be used in place of the different functions to preprocess the FASTA file. It will perform all equivalent functions in one go; the others functions are no longer exported.</li>
<li>
<code>calculate_simpson</code> is used to calculate the Simpson’s index from list of pattern.</li>
<li>
<code>calculate_percent</code> is used to calculate the dissimilarity index from list of pattern.</li>
<li>
<code>check_percent</code> is used to check that the necessary parameter for <code>calculate_percent</code> is passed.</li>
<li>
<code>get_metric_fun</code> is used to translate the metric to the functions, additional metric can be added by defining the functions and overriding the global <code>MinSNPs_metrics</code> variable.</li>
<li>
<code>calculate_simpson</code>, <code>calculate_percent</code>, and <code>check_percent</code>, <code>get_metric_fun</code> are only as default, there is no reason to call these functions directly.</li>
<li>
<code>find_optimised_snps</code> is used to find optimised SNP set(s). It can also be used to recalculate index by giving it included positions and setting max_depth to 0.</li>
<li>
<code>output_result</code> is used to output results from <code>find_optimised_snps</code>.</li>
</ul></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ludwig Kian Soon Hoon, Peter Shaw, Phil Giffard.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

